intel-llvm: init at unstable-2025-11-14

kilyanni opened the pull request:
Adds the Intel LLVM fork for SYCL compilation. This PR includes the compiler as well as an `stdenv` for making nix-derivations that use it.

Some notes:

- Upstreaming the `gnu-install-dirs.patch` patch wont make much sense until the [equivalent vanilla-LLVM patches](https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/compilers/llvm/22/llvm/gnu-install-dirs.patch) are upstreamed (and reach the Intel Fork). The `sycl-jit-exclude-cmake-files.patch` only makes sense with the GNU one already applied, so that neither.
- This is packaged at an unstable commit instead of the last release, because several packaging-related changes are not yet in the release version. Most notably the dependency lookup changes from intel/llvm/issues/19635
- I've successfully used the stdenv to build large SYCL codebases, such as `oneDNN`. As far as I can tell, everything seems to work as intended.

Closes #245031
Related to #367722

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform:
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- Tested, as applicable:
  - [ ] [NixOS tests] in [nixos/tests].
  - [ ] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [ ] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [x] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [ ] Module update: when the change is significant.
- [x] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

kilyanni committed:
intel-llvm: init at unstable-2025-11-14

blenderfreaky commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 470035`
Commit: `87221ed18bcd8197bd02d8eb1e9ff2aac7fb429c`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>intel-llvm</li>
  </ul>
</details>

kuflierl commented:
Finally we actually have intel-llvm. I can not even begin with how many hours I have tried to get the build system to handle offline build resources by reading through those CMake files. Thanks for getting this done (and partially upstreamed).

I will review this when I have a bit of time.

kuflierl commented on pkgs/by-name/in/intel-llvm/unified-runtime.nix:
```diff
@@ -0,0 +1,167 @@
+in
+stdenv.mkDerivation (finalAttrs: {
+  name = "unified-runtime";
+  inherit version;
```
version isn't used anyware else, why is it in a let expr?
```suggestion
  rocmtoolkit_joined = symlinkJoin {
    name = "rocm-merged";

    paths = with rocmPackages; [
      clr
      rocm-comgr
      hsakmt
    ];
  };
in
stdenv.mkDerivation (finalAttrs: {
  name = "unified-runtime";
  version = "0.12.0";
```

blenderfreaky replied:
Good catch, that's probably a left-over from an earlier attempt ^^'

kuflierl reviewed (changes requested):
Although i couldn't look into it deeply (especially gnu-install-dirs.patch), it seems everything looks fine (and works fine according to a few tests) except for a couple of nits.

What errors do you really experience with the built in tests?

Also you will need to squash those commits.

blenderfreaky commented:
I'm gonna have to start a fresh build to tell you what tests exactly fail, but it was mostly (if not entirely) tests in Clang.
All the LLVM tests passed. It's just because of the weird trick with the `ccWrapperStub` that the tests get a clang that behaves slightly differently than they'd expect (and there're many tests that care about those edge cases).

I tried a bunch of different ways to get `lit` to use the unwrapped clang, but couldn't ever get it to work properly.
The simplest way around that would probably be to not build `libdevice` alongside the rest, but I'm not sure it supports standalone builds (as most other Intel-authored LLVM packages don't either)

Also I'm gonna squash everything once the review is done, just keeping it for now to avoid the force-push mess

blenderfreaky commented on pkgs/by-name/in/intel-llvm/unwrapped.nix:
```diff
@@ -0,0 +1,350 @@
+  version = "unstable-2025-11-14";
+  date = "20251114";
+  llvmMajorVersion = "22";
+  src = fetchFromGitHub {
```
This is actually somewhat non-trivial, as `unified-runtime` a bit below in the let block also depends on src.
It could all be moved inside, but even then overriding would still be finnicky because of the wrapping. I suppose something could probably be done via `makeScope` or `makeOverrideable`?

The vanilla LLVM derivation has `src` and co passed in as an input inside the derivation, something similar could work here maybe?

blenderfreaky commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 470035 --additional-package pkgsRocm.intel-llvm`
Commit: `f4f6c98b51f9e2ca30902889badfa05ff3963d7e`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 2 packages built:</summary>
  <ul>
    <li>intel-llvm</li>
    <li>pkgsRocm.intel-llvm</li>
  </ul>
</details>

kuflierl commented:
You accidentally did a typo in your last commit "refator"  ^_^

kuflierl reviewed (approved):
Save for that nit I would say LGTM

blenderfreaky commented:
Squashed everything into one commit, thanks for all the suggestions!

blenderfreaky commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 470035 --additional-package pkgsCuda.intel-llvm --additional-package pkgsRocm.intel-llvm`
Commit: `f7ca69b03186270ccadbcfab317d9e10a342fcb8`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 3 packages built:</summary>
  <ul>
    <li>intel-llvm</li>
    <li>pkgsCuda.intel-llvm</li>
    <li>pkgsRocm.intel-llvm</li>
  </ul>
</details>

RossComputerGuy commented on pkgs/by-name/in/intel-llvm/unwrapped.nix:
```diff
@@ -0,0 +1,331 @@
+  libedit,
+  llvmPackages,
+  # Ideally this would be the LLVM with the same major version
+  # as the LLVM built here (22). However LLVM 22 is packaged
```
LLVM 22 exists as it's own set as of this week.

RossComputerGuy commented on pkgs/by-name/in/intel-llvm/unwrapped.nix:
```diff
@@ -0,0 +1,331 @@
+    parallel-hashmap
+    pkgsStatic.zstd
+  ]
+  ++ unified-runtime.buildInputs;
```
If there's dependencies needed from that, they should be in unified-runtime's `propagatedBuildInputs`. Then you'll only need to add unified-runtime to the build inputs here.

doronbehar replied:
> `unified-runtime` is a bit of a pseudo-package; it's somewhat independent but not really. Their LLVM CMake doesn't allow linking to an externally built UR, instead it's built in-tree (which is why `unified-runtime.cmakeFlags` also gets appended for example).
> 
> The reason I gave it a separate derivation is mostly for separation of concerns, as most of the GPU-driver related code is in there, which is whats most likely to break during any updates (both to Intel LLVM and driver packaging).
> 
> I'm not sure if there's any external users of UR at the moment

I guess this is a fine argument for the meantime, but please consider maybe dismantling `unified-runtime` in the future.

RossComputerGuy reviewed (dismissed):
This is a decent start. I took a cursory glance and have some changes. I do heavily recommend using our existing LLVM infrastructure in nixpkgs to assist.

kuflierl commented:
> Do we really need more, vendor specific LLVMs?

Yes, it would be nice if we could get Intel specific GPU code running. Especially for blender. 

> Why can't they upstream this?

Don't quote me on this but GPU stuff generally isn't upstreamed due to gpus being super vendor specific. Ask upstream.

blenderfreaky commented on pkgs/by-name/in/intel-llvm/unwrapped.nix:
```diff
@@ -0,0 +1,331 @@
+    hwloc
+    emhash
+    parallel-hashmap
+    pkgsStatic.zstd
```
Their CMake explicitly pulls in the static version. Not sure why they do that, but it means (barring patching) it won't make it past configure without.
Also should be noted that the build needs to output static libraries currently due to upstream issues: https://github.com/intel/llvm/issues/19060

doronbehar replied:
@kilyanni using `pkgsStatic` here will cause a mixture of musl & GNU stdlib derivations, which is considered dangerous (so I heard in the past). Have you considered using an overriden derivation of `zstd` with `--disable-shared` and `--enable-static` or something alike?

kilyanni replied:
Ah, I wasn't aware that that pulls in musl
I think `zstd.override { static = true; }` should achieve the same effect as far as this package is concerned

Running the full build with it right now, it passed configure without issues at the least :+1:

kilyanni committed:
unified-memory-framework: fixup static linking

blenderfreaky commented on pkgs/by-name/in/intel-llvm/unwrapped.nix:
```diff
@@ -0,0 +1,331 @@
+in
+# Tip: This build plays nice with ccacheStdenv.
+#      Replace stdenv here to make debugging less tedious.
+stdenv.mkDerivation (finalAttrs: {
```
I think it might be possible, but their current CMake & runtime rather strongly assumes monolithic builds

What exactly are the benefits of splitting here, btw?

In the case of this LLVM fork, you rarely (if ever) want to pull in specific parts, you basically only care about their `clang` for the `-fsycl` support. Basically everything else in it is just a support library for their `clang`, and I'm not aware of any packages that depend on their libllvm or similar

RossComputerGuy replied:
Splitting is good since it helps with cross compilation. It also makes building multiple derivations a bit nicer. There's also the reduction of people's closures.

You can also leverage the existing LLVM infra which makes the size of this PR smaller and more maintainable.

kilyanni commented:
The core changes needed are mostly the PRs by sarnex from https://github.com/intel/llvm/issues/19635.
These do not seem to have been merged into 6.3.0 (their main `sycl` branch and releases seem to be rather divergent).

These changes affect the way dependencies are looked up and thus change the derivation quite a bit.

I suppose I could cherry pick the specific PRs ontop of a release?

Tangentially; as UMF recently got a new release (#493489) and as did L0, I could update this to a newer unstable version now. Though it seems the bleeding edge currently breaks some builds (e.g. due to them removing `syclcompat`)

But either way, merging is blocked on RossComputerGuy at the moment (see the comment chain from the review above, I wrote a small breakdown yesterday regarding some issues with the requested changes)

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-in-distress/3604/131

doronbehar commented on pkgs/by-name/in/intel-llvm/unwrapped.nix:
```diff
@@ -0,0 +1,325 @@
+
+    inherit llvmMajorVersion;
+
+    updateScript = nix-update-script { };
```
FYI, this does nothing, see:

- https://github.com/NixOS/nixpkgs/pull/502056

kilyanni replied:
Oh wow, really? I never knew ^^'

Thanks for the tip!

doronbehar commented on pkgs/by-name/in/intel-llvm/unwrapped.nix:
```diff
@@ -0,0 +1,325 @@
+
+  passthru = {
+    isClang = true;
+
+    inherit unified-runtime;
+
+    # This is for easily referencing version-compatible LLVM libraries
+    # and tools that aren't built in this derivation,
+    # as well as nix tooling, such as the stdenv.
+    baseLlvm = llvmPackages_22;
+
+    inherit llvmMajorVersion;
+
+    updateScript = nix-update-script { };
+
+    # This hardening option causes compilation errors when compiling for amdgcn, spirv and others
+    hardeningUnsupportedFlags = [ "zerocallusedregs" ];
+  };
```
Most people are used to having `meta` be the last argument of `stdenv.mkDerivation`, and `passthru` above it.

doronbehar commented on pkgs/by-name/in/intel-llvm/unwrapped.nix:
```diff
@@ -0,0 +1,325 @@
+    "out"
+    "lib"
+    "dev"
+    "python"
```
```suggestion
    "python"
    # Not adding (conditionally) a "man" output here to avoid complexity in the
    # wrapper, that is not worth it for a few Kb.
```

doronbehar reviewed (commented):
Mostly looks good I guess, but I skimmed over most of the comments. @RossComputerGuy's comments were good so it is a shame they stopped replying and abandoned the review of this PR.

kilyanni commented:
^ the `unified-memory-framework` commit fixes something that broke between updates and only affected the version used by this derivation (with `.override { useJemalloc = false; }`)

kilyanni commented:
Fixed the vet failure and some issues that popped up from rebasing, also squashed all the commits down

All builds pass locally (~~with ccacheStdenv, 'proper' build of all 3 variants would take like 6h or something~~)

@doronbehar I addressed the nits and replaced `pkgsStatic.zstd` with `zstd.override { statice = true; }`. Can you take a look and see if this addresses your issues?

kilyanni commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 470035 --additional-package pkgsRocm.intel-llvm --additional-package pkgsCuda.intel-llvm`
Commit: `635d93e5d440e91fd4836948e5b1c817354d7ae1`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 4 packages built:</summary>
  <ul>
    <li>intel-llvm</li>
    <li>pkgsCuda.intel-llvm</li>
    <li>pkgsRocm.intel-llvm</li>
    <li>unified-memory-framework</li>
  </ul>
</details>

doronbehar commented on pkgs/by-name/in/intel-llvm/unwrapped.nix:
```diff
@@ -0,0 +1,328 @@
+  ++ unified-runtime.buildInputs;
+
+  propagatedBuildInputs = [
+    zstd
```
Shouldn't this be `(zstd.override { static = true; })`? If so, maybe create a `zstdStatic` above under `let`-`in`. I wonder in general: Why do you need to propagate it?

kilyanni replied:
Nice catch ^^

I'm actually not fully sure why I propagated it. I think it had to do with me trying to use the non-static version in an earlier version, then forgetting to remove it.

The build works fine with (static) `zstd` only in `buildInputs`, so I removed the other two

doronbehar commented on pkgs/by-name/in/intel-llvm/unwrapped.nix:
```diff
@@ -0,0 +1,328 @@
+    (lib.cmakeBool "SYCL_INCLUDE_TESTS" false)
+
+    (lib.cmakeFeature "LLVM_ENABLE_ZSTD" "FORCE_ON")
+    (lib.cmakeBool "LLVM_USE_STATIC_ZSTD" true)
```
I assume that using regular `zstd` here and setting it to `false` is something you tested, and that doesn't work. Could you maybe expand on that in a comment?

kilyanni replied:
It seems this is actually already set automatically by some of Intels CMake, so it's redundant to set manually. I removed the explicit `LLVM_USE_STATIC_ZSTD`.

Now `zstd` only appears in `buildInputs`, so I slightly expanded the comment there to explain

doronbehar reviewed (commented):
Looking better, thanks for the corporation! 2 more small comments.

doronbehar reviewed (approved):
Looks good to me. I trust the author to have tested this, and thanks to the other positive reviews I think this should be good enough to go in.

kilyanni commented:
Thanks!
